# Analyzing Traffic Accidents in Bogotá from 2015 to 2021 - Team 82
*José Armando Delgado Álvarez, Andrés Felipe Guzmán Romero, Luis Camilo Jiménez Álvarez, Juan Pablo Lancheros Rodríguez, Paola Andrea Matheus Arbeláez, Alejandro Moreno Fresneda, Leidy Milena Nieves Mendoza, Andrés Felipe Pérez Osorio*

# Data Cleaning

## Importing neccesary libraries
We begin by importing the necessary libraries required to perform our data cleaning and exploratory data analysis

In [ ]:
import numpy                 as np
import pandas                as pd
from pandas.api.types        import CategoricalDtype
import geopandas             as gpd


During this project we will use data stored in the `historico_siniestros_bogota.csv` file containing information on geospatial referenced road accidents in Bogotá from 2015 to 2021.

### Loading and inspecting the data
We read the dataset and display the top rows, in order to have a preliminary view of the data.

In [ ]:
df = pd.read_csv ('data/historico_siniestros_bogota.csv', delimiter=',')
df

X         Y  OBJECTID  FORMULARIO  CODIGO_ACCIDENTE  \
0      -74.090924  4.693807         1  A000640275           4484660   
1      -74.121000  4.603000         2  A001233353          10533499   
2      -74.042000  4.682000         4  A001232786          10533629   
3      -74.166937  4.587187         7  A000200705           4412699   
4      -74.092901  4.607648         8  A000402862           4447845   
...           ...       ...       ...         ...               ...   
199141 -74.160000  4.637000    421907  A001341297          10548522   
199142 -74.167000  4.628000    421908  A001305748          10546116   
199143 -74.158247  4.624830    421909  A001238302          10536074   
199144 -74.167000  4.622000    421910  A001297106          10538181   
199145 -74.168000  4.630000    421911  A001304271          10544226   

          FECHA_OCURRENCIA_ACC  ANO_OCURRENCIA_ACC  \
0       2017/06/12 00:00:00+00                2017   
1       2020/11/19 00:00:00+00                2020   
2       2020/11/10 00:00:00+00                2020   
3       2015/05/11 00:00:00+00                2015   
4       2016/06/08 00:00:00+00                2016   
...                        ...                 ...   
199141  2021/08/30 00:00:00+00                2021   
199142  2021/08/03 00:00:00+00                2021   
199143  2021/03/19 00:00:00+00                2021   
199144  2021/04/18 00:00:00+00                2021   
199145  2021/07/12 00:00:00+00                2021   

                                      DIRECCION     GRAVEDAD  \
0                    AV AVENIDA BOYACA-CL 79 02   SOLO DANOS   
1                             CL 26 S- KR 50 02  CON HERIDOS   
2                              KR 9 - CL 100 02   SOLO DANOS   
3                             CL 63A-KR 72 S 02   SOLO DANOS   
4                                 KR 27-CL 9 14   SOLO DANOS   
...                                         ...          ...   
199141                   KR 86 F  - CL 33  S 02   SOLO DANOS   
199142                  CL 42 B  S- KR 81 L  02  CON HERIDOS   
199143                         DG 2 S- KR 79 12  CON HERIDOS   
199144                        CL 43 S- KR 80 02  CON HERIDOS   
199145  AV AVENIDA CIUDAD DE CALI - CL 42  S 02  CON HERIDOS   

                CLASE_ACC       LOCALIDAD          FECHA_HORA_ACC   LATITUD  \
0                  CHOQUE        ENGATIVA  2017/06/12 05:30:00+00  4.693807   
1                    OTRO   PUENTE ARANDA  2020/11/19 02:05:00+00  4.603000   
2                  CHOQUE         USAQUEN  2020/11/10 13:30:00+00  4.682000   
3                  CHOQUE  CIUDAD BOLIVAR  2015/05/11 10:50:00+00  4.587187   
4                  CHOQUE    LOS MARTIRES  2016/06/08 21:30:00+00  4.607648   
...                   ...             ...                     ...       ...   
199141             CHOQUE         KENNEDY  2021/08/30 16:31:00+00  4.637000   
199142          ATROPELLO         KENNEDY  2021/08/03 14:00:00+00  4.628000   
199143             CHOQUE         KENNEDY  2021/03/19 12:50:00+00  4.624830   
199144             CHOQUE         KENNEDY  2021/04/18 21:21:00+00  4.622000   
199145  CAIDA DE OCUPANTE         KENNEDY  2021/07/12 19:53:00+00  4.630000   

         LONGITUD         CIV  PK_CALZADA  
0      -74.090924  10006772.0    221236.0  
1      -74.121000  16004560.0         NaN  
2      -74.042000  30001107.0         NaN  
3      -74.166937  19001483.0    136166.0  
4      -74.092901  14000548.0    239719.0  
...           ...         ...         ...  
199141 -74.160000   8003090.0         NaN  
199142 -74.167000   8005066.0         NaN  
199143 -74.158247   8005839.0         NaN  
199144 -74.167000   8011660.0         NaN  
199145 -74.168000   8004623.0         NaN  

[199146 rows x 16 columns]

Our dataset has 199,146 rows and 16 columns. Below is a brief description of each column.

|     Variable name    |                    Variable description                            |          Data Type        |
|:--------------------:|:------------------------------------------------------------------:|:-------------------------:|
|X                     |Latitude of the location where the accident occurred                |float64                    |
|Y                     |Longitude of the location where the accident occurred               |float64                    |
|OBJECTID              |Unique identifier for the accident record                           |int64                      |
|FORMULARIO            |Unique alphanumeric identifier for the accident record              |object                     |
|CODIGO_ACCIDENTE      |Code of the accident	                                            |int64                      |
|FECHA_OCURRENCIA_ACC  |Date when the accident occurred in YYYY/MM/DD format                |object                     |
|ANO_OCURRENCIA_ACC    |Year when the accident occurred                                     |int64                      |
|DIRECCION             |Address where the accident occurred                                 |object                     |
|GRAVEDAD              |Severity of the accident occured (3 categories)                     |object                     |
|CLASE_ACC             |Class of the accident occured (7 categories)                        |object                     |
|LOCALIDAD             |Borough of the accident where the accident occurred (20 categories) |object                     |
|FECHA_HORA_ACC        |Time when the accident occurred in HH:MM:SS + 00 format             |object                     |
|LATITUD               |Latitude of the location where the accident occurred                |float64                    |
|LONGITUD              |Longitude of the location where the accident occurred               |float64                    |
|CIV                   |Road Identification Code. See details [here]( 	https://movilidadbogota.maps.arcgis.com/apps/webappviewer/index.html?id=219e91e4397f4f8d99c4c1aaa66ba2c4)              |float64                    |
|PK_CALZADA            |Roadway or carriageway identifier                                   |float64                    |


We also check for missing values across columns

In [ ]:
df.isnull().any()

X                       False
Y                       False
OBJECTID                False
FORMULARIO              False
CODIGO_ACCIDENTE        False
FECHA_OCURRENCIA_ACC    False
ANO_OCURRENCIA_ACC      False
DIRECCION               False
GRAVEDAD                False
CLASE_ACC               False
LOCALIDAD                True
FECHA_HORA_ACC          False
LATITUD                 False
LONGITUD                False
CIV                      True
PK_CALZADA               True
dtype: bool

## Relevance
In order to guarantee that our dataset only contains relevant information for our analysis we:
1. Remove unnecesary variables or observations as follows:
* Columns `X` and `Y` contain the same information as the columns `LONGITUD` and `LATITUD` respectively. Therefore, we remove them from the dataset.
* Columns `FECHA_OCURRENCIA_ACC` and `ANO_OCURRENCIA_ACC` contain the same information in `FECHA_HORA_ACC`
2. Dealing with missing values in the columns `CIV`, `PK_CALZADA`. and `LOCALIDAD`

In [ ]:
#Drop X and Y columns
df = df.drop(["X", "Y"], axis=1)

In [ ]:
#Drop FECHA_OCURRENCIA_ACC and ANO_OCURRENCIA_ACC
df = df.drop(["FECHA_OCURRENCIA_ACC","ANO_OCURRENCIA_ACC"], axis=1)

In [ ]:
#Replace each missing row in CIV and PK_CALZADA with the "Not Available" string
df["CIV"] = df["CIV"].astype(str)
df["CIV"] = df["CIV"].fillna("Not Available")
df["PK_CALZADA"] = df["PK_CALZADA"].astype(str)
df["PK_CALZADA"] = df["PK_CALZADA"].fillna("Not Available")

In [ ]:
#Drop missing values in LOCALIDAD
df = df[df["LOCALIDAD"].notna()]

## Consistency
To achieve consistency in our dataset, we need to:
1. Make sure that all features are represented in the right data type.
* Convert GRAVEDAD, CLASE_ACC and LOCALIDAD to categorical data type
* Convert FECHA_HORA_ACC to datetime
2. Rename and standardize column names
3. Provide appropiate replacements for categories in categorical columns

In [ ]:
#Convert categorical columns to appropriate data types
df["GRAVEDAD"] = df["GRAVEDAD"].astype("category")
df["CLASE_ACC"] = df["CLASE_ACC"].astype("category")
df["LOCALIDAD"] = df["LOCALIDAD"].astype("category")

In [ ]:
#Convert FECHA_HORA_ACC to datetime
df["FECHA_HORA_ACC"] = pd.to_datetime(df["FECHA_HORA_ACC"])

In [ ]:
#Rename columns with lowercase and underscore format
columns_dict = {"OBJECTID":"id", 
                "FORMULARIO":"form_id",
                "CODIGO_ACCIDENTE": "accident_code",
                "DIRECCION":"address",
                "GRAVEDAD":"severity",
                "CLASE_ACC":"accident_type",
                "LOCALIDAD":"borough",
                "FECHA_HORA_ACC":"full_date",
                "LATITUD":"latitude",
                "LONGITUD":"longitude",
                "CIV":"civ",
                "PK_CALZADA":"road_id",
               }
df = df.rename(columns = columns_dict)

In [ ]:
#severity column categories
rename_dict_severity = {"SOLO DANOS": "Only damages", "CON HERIDOS":"With Injured", "CON MUERTOS":"With deceased"}
df["severity"] = df["severity"].cat.rename_categories(rename_dict_severity)

In [ ]:
#accident type column categories
rename_dict_acctype = {"ATROPELLO":"Run-over", "AUTOLESION":"Self-harm",
                        "CAIDA DE OCUPANTE":"Passenger falling", "CHOQUE":"Crash", 
                        "INCENDIO":"Fire","OTRO":"Other", "VOLCAMIENTO":"Overturn"}
df["accident_type"] = df["accident_type"].cat.rename_categories(rename_dict_acctype)

In [ ]:
#borough column categories (Antonio Nariño)
rename_dict_boroughs = {"ANTONIO NARINO": "ANTONIO NARIÑO"}
df["borough"] = df["borough"].cat.rename_categories(rename_dict_boroughs)

## Data Augmentation
1. Creating new features in the dataset:
* year column
* month column
* day_of_week column
* hour column
2. Merging our dataset with a geojson dataset containing polygons of boroughs to perform geospatial analysis


In [ ]:
#We extract the year, month, day name, and hour from the "date" column, store them in separate columns and rearrange the resulting categories in a chronological order
#Year
df["year"] = df["full_date"].dt.year
#Month
df["month"] = df["full_date"].dt.month_name()
mon = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
df['month']= pd.Categorical(df['month'], categories=mon, ordered=True)
#Weekday
df["day_of_week"] = df["full_date"].dt.day_name()
wday = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
df['day_of_week']= pd.Categorical(df['day_of_week'], categories=wday, ordered=True)
#Hour
df["hour"] = df["full_date"].dt.hour
#Month and Year
df["month_year"] = df["month"].astype(str).str[:3] + ", " + df["year"].astype(str)

In [ ]:
#Export the clean dataset as df_clean.csv
df.to_csv("data/df_clean.csv", encoding = "utf-8", index = False)

### Visualizing the cleaned dataset

In [ ]:
df.head()

id     form_id  accident_code                     address      severity  \
0   1  A000640275        4484660  AV AVENIDA BOYACA-CL 79 02  Only damages   
1   2  A001233353       10533499           CL 26 S- KR 50 02  With Injured   
2   4  A001232786       10533629            KR 9 - CL 100 02  Only damages   
3   7  A000200705        4412699           CL 63A-KR 72 S 02  Only damages   
4   8  A000402862        4447845               KR 27-CL 9 14  Only damages   

  accident_type         borough                 full_date  latitude  \
0         Crash        ENGATIVA 2017-06-12 05:30:00+00:00  4.693807   
1         Other   PUENTE ARANDA 2020-11-19 02:05:00+00:00  4.603000   
2         Crash         USAQUEN 2020-11-10 13:30:00+00:00  4.682000   
3         Crash  CIUDAD BOLIVAR 2015-05-11 10:50:00+00:00  4.587187   
4         Crash    LOS MARTIRES 2016-06-08 21:30:00+00:00  4.607648   

   longitude         civ   road_id  year     month day_of_week  hour  \
0 -74.090924  10006772.0  221236.0  2017      June      Monday     5   
1 -74.121000  16004560.0       nan  2020  November    Thursday     2   
2 -74.042000  30001107.0       nan  2020  November     Tuesday    13   
3 -74.166937  19001483.0  136166.0  2015       May      Monday    10   
4 -74.092901  14000548.0  239719.0  2016      June   Wednesday    21   

  month_year  
0  Jun, 2017  
1  Nov, 2020  
2  Nov, 2020  
3  May, 2015  
4  Jun, 2016

### Merging the Geographic Dataset with the Road Accidents Dataset

In [ ]:
# Loading the shape file for Bogotá polygons of boroughs
shp_df = gpd.read_file('data/localidades/localidades.shp')
# Cleaning the dataframe before merging
shp_df = shp_df.drop(["OBJECTID","CODIGO_LOC","DECRETO","LINK","SIMBOLO","ESCALA_CAP","FECHA_CAPT", "SHAPE_LEN"], axis=1)
geo_columns_dict = {"NOMBRE":"borough", "SHAPE_AREA":"shape_area"}
shp_df = shp_df.rename(columns = geo_columns_dict)
shp_df.to_csv("data/geospatial.csv")

In [ ]:
#Merging accident counts and shapefile
df_borough = df.groupby(['borough', 'year']).size().reset_index(name="accident_count")
geospatial_df = shp_df.merge(df_borough, how="left", left_on=['borough'], right_on=['borough'])

In [ ]:
geospatial_df.head()

borough    shape_area                                           geometry  \
0  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   
1  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   
2  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   
3  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   
4  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   

   year  accident_count  
0  2015             603  
1  2016             813  
2  2017             966  
3  2018             993  
4  2019             991

### Merging the Population Dataset with the Road Accidents Dataset

Importing the population dataset

In [ ]:
df_pop = pd.read_excel('data/poblacion_localidades.xlsx',sheet_name=1, skiprows=5)
#Remove Total Population for Bogotá
df_pop = df_pop.drop(df_pop.index[20:22])
#Drop Código Localidad and column years that are not in the main dataset (We are only interested in population from year 2015 to year 2021)
df_pop.drop(df_pop.columns[[0, 2,3,4,5,6,7,8,9,10,11,19,20,21,22,23,24,25,26,27,28,29,30,31,32]], axis = 1, inplace = True)
#Change Nombre de Localidad column name
pop_columns_dict = {"Nombre localidad":"borough"}
df_pop = df_pop.rename(columns = pop_columns_dict)
#From Wide to Tidy Format
df_pop = df_pop.melt(id_vars=["borough"], var_name="year", value_name="population")
df_pop

borough  year  population
0               USAQUEN  2015    520480.0
1             CHAPINERO  2015    151996.0
2              SANTA FE  2015    102155.0
3         SAN CRISTOBAL  2015    382247.0
4                  USME  2015    353558.0
..                  ...   ...         ...
135       PUENTE ARANDA  2021    253367.0
136       LA CANDELARIA  2021     17877.0
137  RAFAEL URIBE URIBE  2021    383960.0
138      CIUDAD BOLIVAR  2021    649834.0
139             SUMAPAZ  2021      3584.0

[140 rows x 3 columns]

In [ ]:
df_merged = pd.merge(geospatial_df, df_pop, on=['borough','year'], how='left')
df_merged.head(3)

borough    shape_area                                           geometry  \
0  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   
1  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   
2  SANTA FE  4.517065e+07  POLYGON ((100996.362 103506.019, 101013.606 10...   

   year  accident_count  population  
0  2015             603    102155.0  
1  2016             813    102485.0  
2  2017             966    102984.0

## Attribution
* "Road accidents in Bogotá from 2015 to 2021", May 20, 2022. Datos Abiertos Bogotá created by [Secretaria Distrital de Movilidad](https://www.movilidadbogota.gov.co/web/datos_abiertos),  available at: https://datosabiertos.bogota.gov.co/en/dataset/historico-siniestros-bogota-d-c
* "Boroughs in Bogotá D.C", June 6, 2022. SERIES. Servicios de Información Espacial, available at: https://sites.google.com/site/seriescol/shapes
* "Bogotá demographic data per borough from 2005 to 2035", June 6, 2022. Visor de Población - Secretaría Distrital de Planeación, available at https://sdpbogota.maps.arcgis.com/apps/MapSeries/index.html?appid=baabe888c3ab42c6bb3d10d4eaa993c5